### PR TITLE
Us#82

### PR DESCRIPTION
**Pull request by prabhanjan2906**
**Taiga User Story**
TG-82

**Taiga Task(s)**
TG-119 

**What did you change?**
Introduced the retrieval of the data of all the patients that belong into the particular trial into CSV.
Changed files:
1. Added controller/handler/trial-csv.js
2. view/template/trial.hbs - Added a new Button to retrieve the data into CSV.
3. controller/router.js - Added a new route to get the trial id and forward to trial-csv.

**How can we test?**
In the Main Dashboard file, goto any trial. You will see a new Button called "Export Trial Data". Click on that button and you should see a csv file getting downloaded. This file contains the answers to the questions each patient answered.

### CODE REVIEW

File: controller/handler/trial-csv.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    )
+    .then((optionsWithAnswers) => {
+        const property = ['pin', 'name', 'id', 'questionText', 'questionId'];
+        const uniqueAnswers = deduplicate(optionsWithAnswers, property);
+
+        return convertJsonToCsv(uniqueAnswers, configuration);
+    })
+    .then((csv) => {
+        reply(csv).type('text/csv');
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: `return` reply here

File: controller/handler/trial-csv.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        key: 'questionText'
+    },
+    {
+        label: 'question option',
+        key: 'optionText'
+    }
+];
+
+/**
+ * Create a Comma Seperate Value export of a single patient's data.
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: This exports all patients for a trial

File: controller/handler/trial-csv.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * @param {Request} request - Hapi request
+ * @param {Reply} reply - Hapi Reply
+ * @returns {View} Rendered page
+ */
+function trialCSV (request, reply) {
+    database.sequelize.query(
+        `
+        SELECT pa.pin, st.name, si.id, qt.id AS questionId, qt.questionText, qo.id AS optionId, qo.optionText
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: we may want patient more details about the patient to be exported. (such as start, end, and device used)

Thoughts @promisasu/admin ?

File: controller/handler/trial-csv.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        ON jsq.surveyTemplateId = st.id
+        JOIN question_template AS qt
+        ON qt.id = jsq.questionTemplateId
+        JOIN question_option AS qo
+        ON qo.questionTemplateId = qt.id
+        JOIN question_result AS qr
+        ON qr.surveyInstanceId = si.id
+        AND qr.questionOptionId = qo.id
+        WHERE tr.id = ?
+                and si.state = 'completed'
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: SQL keywords should be in all caps. (in this case `AND`)

File: controller/handler/trial-csv.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                ORDER BY si.id, jsq.questionOrder, qo.order
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: lines 68 and 69 should have the same indent as the rest of the query

File: controller/handler/trial-csv.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        `,
+        {
+            type: database.sequelize.QueryTypes.SELECT,
+            replacements: [
+                request.params.id
+            ]
+        }
+    )
+    .then((optionsWithAnswers) => {
+        const property = ['pin', 'name', 'id', 'questionText', 'questionId'];
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: Device info, patient start and end should be deduplicated here.

File: controller/handler/trial-csv.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * @module controller/handler/trial-csv
+ */
+
+const database = require('../../model');
+const convertJsonToCsv = require('../helper/convert-json-to-csv');
+const boom = require('boom');
+const deduplicate = require('../helper/deduplicate');
+const configuration = [
+    {
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: Should trial name be included here? @poojaRal @iamaureen

File: controller/handler/trial-csv.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+ * @param {Reply} reply - Hapi Reply
+ * @returns {View} Rendered page
+ */
+function trialCSV (request, reply) {
+    database.sequelize.query(
+        `
+        SELECT pa.pin, pa.deviceType, pa.deviceVersion, pa.dateStarted, pa.dateCompleted, st.name,
+        si.id, qt.id AS questionId, qt.questionText, qo.id AS optionId, qo.optionText
+        FROM trial AS tr
+        JOIN stage AS stage
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: `JOIN stage AS stage`
Should be either
`JOIN stage AS st`
Or
`JOIN stage`

**Comment by ChristianMurphy** [comment]
There are merge conflicts between this and development that need to be resolved

File: controller/handler/trial-csv.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+ */
+function trialCSV (request, reply) {
+    database.sequelize.query(
+        `
+        SELECT pa.pin, pa.deviceType, pa.deviceVersion, pa.dateStarted, pa.dateCompleted, st.name,
+        si.id, qt.id AS questionId, qt.questionText, qo.id AS optionId, qo.optionText
+        FROM trial AS tr
+        JOIN stage
+        ON stage.trialId = tr.id
+        JOIN patient AS pa
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: this query will need to pull from `active_patients` instead of `patient`

**Comment by ChristianMurphy** [comment]
screenshot
Lets set default value for the deduplicator to an empty string, all the 'DNE' and `underfined` values are not fun to look at.

screenshot2
device type, device version, start and stop are all patient related and should be adjacent to the patient information on the table.

**Comment by ChristianMurphy** [comment]
screenshot3
A couple things
The undefined and DNE values are still showing.
and the question prompt and the question answer are separated by the device information, which seems a bit odd.

**Comment by poojaRal** [comment]
Please remove columns : Device Type and Device Version.
Please remove duplicate entries for Start and End Date. Display it only once as the PIN number.

File: controller/handler/trial-csv.js [hunk 10]
```diff
@@ -53,6 +43,16 @@ const configuration = [
         label: 'question option',
         key: 'optionText',
         default: ''
+    },
+    {
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: This is moving the start and end away from the patient information.
It should be next to the patient pin number.

**Comment by poojaRal** [comment]
The duplicate data of dates has been removed. 

The csv generated looks like below from your code -

[current.xlsx](https://github.com/promisasu/Portal/files/237634/current.xlsx)

As per my understanding what Dr. Gary wanted that for each PIN, all survey instances for that PIN must be listed under that PIN once.

For eg : 
[2.xlsx](https://github.com/promisasu/Portal/files/237629/2.xlsx)

I have attached an example excel format of the way the csv should look like.
Here you can see that for one PIN : 2001 if there are two survey instances then both should be listed one after another before we jump onto PIN : 2002 displaying its survey instances.

@kgary, @iamaureen and @ChristianMurphy : Please check and review.
Please comment on what should be a correct csv format.
Please attach sample excel sheets for better clarity of what the requrement is.

**Comment by ChristianMurphy** [comment]
Personally I would lean toward having trial, patient and survey data columns grouped.
And avoiding extra rows for trial and patient.

Like so:

image

**Comment by kgary** [comment]
I agree the data should be de-normalized in this way. WHy is there no question prompt for the 1st question?

**Comment by poojaRal** [comment]
1st Question (in our case question_no 31) is a two part question.
However the question text is a single line.
Answers submitted by the user are body pain location and intensity.
Thus we display question text just once for this question and answer options selected by the user are shown on 2 different rows.

**Comment by poojaRal** [comment]
@prabhanjan2906 : I guess now you can look at the excel which Christian Murphy has posted in his comment as a sample. Please refer the same and do the required changes.

File: controller/handler/trial-csv.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+const boom = require('boom');
+const deduplicate = require('../helper/deduplicate');
+const configuration = [
+    {
+        label: 'Trial Name',
+        key: 'trialName',
+        default: ''
+    },
+    {
+        label: 'patient pin',
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: Could you also use the column names from the example posted?

File: controller/handler/trial-csv.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            type: database.sequelize.QueryTypes.SELECT,
+            replacements: [
+                formatSpecifier,
+                formatSpecifier,
+                request.params.id
+            ]
+        }
+    )
+    .then((optionsWithAnswers) => {
+        const property = ['trialName', 'pin', 'name', 'id', 'questionText', 'questionId',
```
Comment (ChristianMurphy) [inline controller/handler/trial-csv.js]: IMO the
``` js
const property = [
     'TrialName',
     'Pin',
     ...
]
```

**Comment by poojaRal** [comment]
LGTM

**Comment by ChristianMurphy** [comment]
LGTM

**Pull request merged**